### ISSUE TITLE
IMPROVED_POM_SUPPORT: Excludes on first-order overwritten by transitive

### ISSUE THREAD

**Issue by sghill** [nebula-plugins/gradle-nebula-integration#12]
Example project: https://github.com/nebula-plugins/gradle-nebula-integration/tree/master/consumer-excludes-with-bom

Given a build file with dependencies declared like this:
```
dependencies {
  implementation('org.springframework:spring-beans:5.0.6.RELEASE') { // depends on snakeyaml 1.20
    exclude group: 'org.yaml', module: 'snakeyaml'
  }
  implementation 'nebulatest:mylib:latest.release' // depends on snakeyaml:1.19 and spring-beans:5.0.6.RELEASE
}
```

I would expect spring-beans opinion of snakeyaml to not participate in conflict resolution since it has been excluded, resulting in 1.19 of snakeyaml.

## Without IMPROVED_POM_SUPPORT
```
$ ./gradlew -DimprovedPom=false -q dependencies --configuration compileClasspath

------------------------------------------------------------
Root project
------------------------------------------------------------

compileClasspath - Compile classpath for source set 'main'.
+--- org.springframework:spring-beans:5.0.6.RELEASE
|    \--- org.springframework:spring-core:5.0.6.RELEASE
|         \--- org.springframework:spring-jcl:5.0.6.RELEASE
\--- nebulatest:mylib:latest.release -> 1.0.1
     +--- org.yaml:snakeyaml:1.19                                                    <-- expected
     \--- org.springframework:spring-beans:5.0.6.RELEASE (*)

(*) - dependencies omitted (listed previously)
```

## With IMPROVED_POM_SUPPORT
```
$ ./gradlew -DimprovedPom=true -q dependencies --configuration compileClasspath

------------------------------------------------------------
Root project
------------------------------------------------------------

compileClasspath - Compile classpath for source set 'main'.
+--- org.springframework:spring-beans:5.0.6.RELEASE
|    +--- org.springframework:spring-core:5.0.6.RELEASE
|    |    \--- org.springframework:spring-jcl:5.0.6.RELEASE
|    \--- org.yaml:snakeyaml:1.20                                                           <-- not expected
\--- nebulatest:mylib:latest.release -> 1.0.1
     +--- org.yaml:snakeyaml:1.19 -> 1.20
     \--- org.springframework:spring-beans:5.0.6.RELEASE (*)

(*) - dependencies omitted (listed previously)
```

**Comment by bigdaz** [comment]
The behaviour you're describing here is a combination of 2 different issues:

- [Gradle processes exclusions differently from Maven](https://github.com/gradle/gradle/issues/1473). If a module can be reached via multiple paths in the dependency graph, it must be excluded from _all_ of these paths in order to be excluded from the dependency result (see https://github.com/gradle/gradle/issues/1473#issuecomment-342938454).  
    - In this case, `nebulatest:mylib:1.0.1` has a dependency on `spring-beans:5.0.6-RELEASE`, and `snakeyaml` is not excluded from this dependency path.
    - This behaviour is independent of `IMPROVED_POM_SUPPORT`
- [Improved pom support gives unexpected importance to versions of optional dependencies](https://github.com/gradle/gradle/issues/5434). One aspect of `IMPROVED_POM_SUPPORT` is support for dependencies marked as `optional` in a POM file. Gradle creates a dependency constraint for each optional dependency, allowing a consumer to add a dependency but omit the version. 
    - The problem of "unexpected importance" is that Gradle creates a regular `require` version constraint for these optional dependencies, which means that Gradle will conflict resolve to choose the highest version from all declared constraints.
    - One possible fix for this is to use weaker `prefer` version constraints for optional dependencies. That would mean that the `optional` version would only be used in the case that no other dependency versions were specified.

We definitely plan on addressing the second issue before releasing Gradle 5.0 and enabling `IMPROVED_POM_SUPPORT` by default. Would this be sufficient to resolve this issue for you?

**Comment by sghill** [comment]
Thanks for the explanation and links @bigdaz! Gradle 5.0 sounds good.

**Comment by bigdaz** [comment]
@sghill Can you confirm that if we fix the second issue I mentioned above that will satisfy your requirements? The first issue (excludes don't work like Maven) is more fundamental, and will be more difficult to fix.

**Comment by sghill** [comment]
@bigdaz yes, confirming I expect only the second issue to be fixed in Gradle 5.0.

**Comment by sghill** [comment]
Gradle 5.0 will not translate optional dependencies into constraints, effectively solving this issue.
